### ISSUE TITLE
Not working with freezeTableName

### ISSUE THREAD

**Issue by fraywing** [DanielHreben/sequelize-transparent-cache#34]
Getting error when using `findByPk`:

```
error: Error: MemberOrganization has not been defined
    at Sequelize.model (/Users/austin/Dev/RupieNetwork/node_modules/sequelize/lib/sequelize.js:371:13)
    at Object.keys.forEach (/Users/austin/Dev/RupieNetwork/node_modules/sequelize-transparent-cache/lib/cache/util.js:46:27)
    at Array.forEach (<anonymous>)
    at loadAssociations (/Users/austin/Dev/RupieNetwork/node_modules/sequelize-transparent-cache/lib/cache/util.js:35:35)
    at dataToInstance (/Users/austin/Dev/RupieNetwork/node_modules/sequelize-transparent-cache/lib/cache/util.js:12:15)
    at client.get.then.data (/Users/austin/Dev/RupieNetwork/node_modules/sequelize-transparent-cache/lib/cache/index.js:60:12)
    at propagateAslWrapper (/Users/austin/Dev/RupieNetwork/node_modules/async-listener/index.js:504:23)
    at /Users/austin/Dev/RupieNetwork/node_modules/async-listener/glue.js:188:31
    at /Users/austin/Dev/RupieNetwork/node_modules/async-listener/index.js:541:70
    at /Users/austin/Dev/RupieNetwork/node_modules/async-listener/glue.js:188:31
```

Traced it down to this block in `utils` line `38`: 

```
  if (model.associations[key].hasOwnProperty('options')) {
      const modelName = model.associations[key].options.name.singular
      associations.push({
        model: model.sequelize.model(modelName),
        as: key
      })
    }
```

My model looks like this: 

```
const OrganizationxRolexProject = sequelizeConfig.define(
  'OrganizationxRolexProject',
  {
    RoleID: Sequelize.INTEGER
  }
);
```

and the association looks like this:

```
   Project.belongsToMany(models.Organization, {
    as: 'MemberOrganizations',
    through: models.OrganizationxRolexProject,
    foreignKey: 'ProjectID',
    otherKey: 'OrganizationID'
  });
```

TL;DR; `model.associations[key].options.name.singular` isn't the right value for the actual model, and only provides the `as` association alias, so `model.sequelize.model(modelName)` throws.

**Comment by DanielHreben** [comment]
Hi @ignacioblit ! Can you take a look?

**Comment by ignacioblit** [comment]
Hi @fraywing thanks for the issue and the observation.

It took me a while to figure out how to dynamically fetch associations at that point of the execution and `model.associations[key].options.name.singular` provided the most 'trusty' way of getting their names and making this work. 

If you know of a better way of making this work I'm open to look into it.

Thanks!

**Comment by fraywing** [comment]
Hmm, not entirely certain, I'll look into it. This does seem to limit this library pretty significantly, however. Since it doesn't play well with Aliases.

**Comment by ignacioblit** [comment]
It would be great if you could do so, i still did not find a workaround. 

Something to take into account is that this limitation only applies when trying to cache joined models. Something which was not possible a few months ago.

**Comment by fraywing** [comment]
I see. So would it make sense to add a config to disable caching includes? We'd still get a performance boost if the main model was cached + the includes were fetched and merged into the cached object.

**Comment by ignacioblit** [comment]
You're right. That would not be a bad idea

**Comment by fraywing** [comment]
Cool, do you want me to try to tackle that? Not familiar with the code, should I just gate the include logic based on config prop?

**Comment by DanielHreben** [comment]
@fraywing @ignacioblit as far as I understand, it's impossible to get real model name having association name using sequelize api. Can we save this meta info into cache as well as data itself?

**Comment by ignacioblit** [comment]
@fraywing please, if you are up to it do it :D

Gating that part of the code based on a config property would solve this issue if you don't mind not loading any associations. 

A broader approach would be what @DanielHreben proposes and save this meta info as part of the cache too, enabling this way to have the correct model names when needed. 

What worries me about this last approach is how to accomplish this without adding overhead to the end programmer. 

Ok... While writing this an idea came to my mind and I did some quick tests, I may have somewhat of a workaround:

`model.associations[key].options.name` has whitin itself the singular and plural properties (which have the names of the model)

`model.associations[key].options.as` seems to have the as (or alias name).

What about doing `model.sequelize.isDefined(modelName)` with these three names until one is found or returning false in the end?

Such approach would not throw, it would increase the cases when this would work and will also make the configuration for the associations unnecessary as it would work either way. 

Cheers!

**Comment by fraywing** [comment]
That's a really good idea, @ignacioblit. I like it. 

Do you have a fork/branch where you have this working?

**Comment by ignacioblit** [comment]
Hey @fraywing thank you! 

Not really I just quickly tested all the ways I could access model names. 

If you are up to it I'm confident you can pull this one off ;)

You already know where the code that needs to be changed is. While working on this you can use the tests in the codebase to check your changes are working as expected. You can also add an Alias test case to easily repeatedly test this while working. 

Let me know if you need any help

**Comment by fraywing** [comment]
shoot, might not be able to look into this due to schedule. I appreciate your advice, and I think this is a solid direction!

**Comment by fraywing** [comment]
Yo friends, ended up fixing it afterall. 

Here is my changed `loadAssociations` method:


```function loadAssociations (model, depth = 1) {
  const associations = [];
  // eslint-disable-next-line
  if (Object.keys(model.associations).length && depth <= 3) {
    Object.entries(model.associations).forEach(([
      key, value
    ]) => {
      // go get the model name
      const target = sequelize.model(value.target.name);
      // we have to do this to get scopes to work
      target._injectScope({});
      // eslint-disable-next-line
        associations.push({
        model: target,
        as: value.associationAccessor,
        // recursive lookup
        include: loadAssociations(target, depth + 1)
      });
    });
  }

  return associations;
}```

I would PR, but I ended up heavily modifying the library to suite our project style needs.

**Comment by DanielHreben** [comment]
Oh wow, that's mad! We should cover this with large amount of unit tests if we wanna include this code to project. I'm working on changing approach to testing to make it more easy to write unit tests and have separate suite of integration tests to make it possible. 
Another option - we can allow users to pass loadAssociations() function as library option, so this code will not be a part of project, but problem will be solvable without forking library.

**Comment by ignacioblit** [comment]
Hey guys I finally came to it because I found the need of the multi level loading using the depth approach @fraywing used.

This is the PR https://github.com/DanielHreben/sequelize-transparent-cache/pull/39

**Comment by DanielHreben** [comment]
Closing this, thanks to all contributors and @p-fedyukovich in particular.
Please feel free to reopen.
